seastar/src/distributed.rs

```diff
@@ -291,11 +312,20 @@ impl<S: Service> Distributed<S> {
         crate::assert_runtime_is_running();
 
         let distr = self._inner.clone();
-        submit_to(shard_id, || async move {
+        let lock = self._locks[shard_id as usize].clone();
+        submit_to(shard_id, move || async move {
+            let lock = lock.try_read();
+            if lock.is_err() {
+                panic!(
+                    "Data race detected: can't lock instance {} for read-access",
```

Comment:
The message is misleading - it's not a data race. Data race happens when two or more threads try to access the same data in an unsynchronized way. Here, the lock will always be taken by only one thread - so it's more like you prevent a violation of Rust's aliasing rules. Perhaps you could make the message more similar to what a `RefCell` generates when it detects a rules violation.

Anyway, a more useful error message would also mention possible reasons for why the panic happened so that the developer quickly understands what they did wrong.